File: crates/piggy-pigpen/src/hyphence.rs
//! Minimal hyphence framing (madder RFC 0001) — the subset pigpen needs.
//!
//! Not a full RFC 0001 implementation: only the `# - @ !` prefixes, the
//! `---\n` boundaries, the required blank-line body separator, and the
//! `@`-XOR-body rule. The cutover (RFC 0008 "Compatibility") replaces
//! this with either piggy's own conforming framing or a shared library.

use crate::{Error, Result};

pub const BOUNDARY: &str = "---\n";

/// One metadata line: a single-byte prefix and its content.
#[derive(Clone)]
pub struct MetaLine {
    pub prefix: u8,
    pub body: String,
}

/// Framing-level view of a document.
#[derive(Default)]
pub struct HyphenceDoc {
    pub meta: Vec<MetaLine>,
    pub body: Vec<u8>,
}

impl HyphenceDoc {
    fn has_at_ref(&self) -> bool {
        self.meta.iter().any(|l| l.prefix == b'@')
    }

    /// Render just the metadata section (both boundaries) in canonical
    /// RFC 0001 order: `#`, `-`, `@`, `!`. Input order preserved within a
    /// prefix.
    pub fn marshal_metadata(&self) -> Vec<u8> {
        let mut out = Vec::new();
        out.extend_from_slice(BOUNDARY.as_bytes());
        for want in [b'#', b'-', b'@', b'!'] {
            for l in &self.meta {
                if l.prefix == want {
                    out.push(l.prefix);
                    out.push(b' ');
                    out.extend_from_slice(l.body.as_bytes());
                    out.push(b'\n');
                }
            }
        }
        out.extend_from_slice(BOUNDARY.as_bytes());
        out
    }

    /// Render the full document.
    pub fn marshal(&self) -> Result<Vec<u8>> {
        if self.has_at_ref() && !self.body.is_empty() {
            return Err(Error::Hyphence(
                "'@' blob reference together with an inline body".into(),
            ));
        }
        let mut out = self.marshal_metadata();
        if !self.body.is_empty() {
            out.push(b'\n');
            out.extend_from_slice(&self.body);
        }
        Ok(out)
    }
}

/// Decode the framing of a pigpen document.
pub fn parse(raw: &[u8]) -> Result<HyphenceDoc> {
    let text = raw;
    let mut pos = 0usize;

    let first = read_line(text, &mut pos).ok_or_else(|| Error::Hyphence("empty input".into()))?;
    if first != BOUNDARY.as_bytes() {
        return Err(Error::Hyphence("missing opening '---' boundary".into()));
    }

    let mut doc = HyphenceDoc::default();
    let mut closed = false;
    while let Some(line) = read_line(text, &mut pos) {
        if line == BOUNDARY.as_bytes() {
            closed = true;
            break;
        }
        let content = strip_lf(line);
        if content.len() < 2 || content[1] != b' ' {
            return Err(Error::Hyphence(format!(
                "malformed metadata line: {:?}",
                String::from_utf8_lossy(line)
            )));
        }
        let prefix = content[0];
        match prefix {
            b'#' | b'-' | b'@' | b'!' => doc.meta.push(MetaLine {
                prefix,
                body: String::from_utf8_lossy(&content[2..]).into_owned(),
            }),
            other => {
                return Err(Error::Hyphence(format!(
                    "unknown metadata prefix {:?}",
                    other as char
                )));
            }
        }
    }
    if !closed {
        return Err(Error::Hyphence("missing closing '---' boundary".into()));
    }

    // Remainder is the body, preceded by the required blank-line separator.
    let rest = &text[pos..];
    if !rest.is_empty() {
        if rest[0] != b'\n' {
            return Err(Error::Hyphence(
                "body present without blank-line separator".into(),
            ));
        }
        doc.body = rest[1..].to_vec();
    }
    if doc.has_at_ref() && !doc.body.is_empty() {
        return Err(Error::Hyphence(
            "'@' blob reference together with an inline body".into(),
        ));
    }
    Ok(doc)
}

fn read_line<'a>(buf: &'a [u8], pos: &mut usize) -> Option<&'a [u8]> {
    if *pos >= buf.len() {
        return None;
    }
    let start = *pos;
    match buf[start..].iter().position(|&b| b == b'\n') {
        Some(i) => {
            *pos = start + i + 1;
            Some(&buf[start..*pos])
        }
        None => {
            *pos = buf.len();
            Some(&buf[start..])
        }
    }
}

fn strip_lf(line: &[u8]) -> &[u8] {
    line.strip_suffix(b"\n").unwrap_or(line)
}

File: go/pigpen/hyphence.go
package pigpen

import (
	"bufio"
	"bytes"
	"errors"
	"fmt"
	"strings"
	"unicode/utf8"
)

// Minimal hyphence framing (madder RFC 0001). Only the subset pigpen
// needs: a metadata section delimited by "---\n" boundaries, the
// prefixes pigpen uses (# - @ !), a required blank-line separator
// before an inline body, and the @-XOR-body rule.
//
// This is intentionally NOT a full RFC 0001 implementation (no % comment
// entanglement, no < standalone object refs, no lenient mode). It is the
// envelope sketch RFC 0008 "Compatibility" describes.

const boundary = "---\n"

var (
	errNoOpeningBoundary   = errors.New("pigpen: missing opening '---' boundary")
	errNoClosingBoundary   = errors.New("pigpen: missing closing '---' boundary")
	errMissingSeparator    = errors.New("pigpen: body present without blank-line separator")
	errAtRefWithInlineBody = errors.New("pigpen: '@' blob reference together with an inline body")
	errUnknownPrefix       = errors.New("pigpen: unknown metadata prefix")
)

// metaLine is one decoded metadata line: a single-byte prefix and its
// content (everything after "<prefix> ", trimmed of the trailing LF).
type metaLine struct {
	prefix byte
	body   string
}

// hyphenceDoc is the framing-level view: ordered metadata lines plus the
// raw body bytes (empty if none).
type hyphenceDoc struct {
	meta []metaLine
	body []byte
}

func (h *hyphenceDoc) hasInlineBody() bool { return len(h.body) > 0 }

func (h *hyphenceDoc) hasAtRef() bool {
	for _, l := range h.meta {
		if l.prefix == '@' {
			return true
		}
	}
	return false
}

// marshalMetadata renders just the metadata section (both boundaries
// included) in canonical RFC 0001 order: descriptions (#), tags/refs (-),
// blob (@), type (!). Within a prefix, input order is preserved.
func (h *hyphenceDoc) marshalMetadata() []byte {
	var b bytes.Buffer
	b.WriteString(boundary)
	for _, want := range []byte{'#', '-', '@', '!'} {
		for _, l := range h.meta {
			if l.prefix == want {
				fmt.Fprintf(&b, "%c %s\n", l.prefix, l.body)
			}
		}
	}
	b.WriteString(boundary)
	return b.Bytes()
}

// marshal renders the full document: metadata section, then (if there is
// an inline body) the blank-line separator and the body bytes.
func (h *hyphenceDoc) marshal() ([]byte, error) {
	if h.hasAtRef() && h.hasInlineBody() {
		return nil, errAtRefWithInlineBody
	}
	out := h.marshalMetadata()
	if h.hasInlineBody() {
		out = append(out, '\n')
		out = append(out, h.body...)
	}
	return out, nil
}

// parseHyphence decodes the framing of a pigpen document.
func parseHyphence(raw []byte) (*hyphenceDoc, error) {
	r := bufio.NewReader(bytes.NewReader(raw))

	first, err := r.ReadString('\n')
	if err != nil || first != boundary {
		return nil, errNoOpeningBoundary
	}

	doc := &hyphenceDoc{}
	closed := false
	for {
		line, err := r.ReadString('\n')
		if err != nil {
			return nil, errNoClosingBoundary
		}
		if line == boundary {
			closed = true
			break
		}
		// "<prefix> <content>\n"
		content := strings.TrimSuffix(line, "\n")
		if len(content) < 2 || content[1] != ' ' {
			return nil, fmt.Errorf("pigpen: malformed metadata line: %q", line)
		}
		prefix := content[0]
		switch prefix {
		case '#', '-', '@', '!':
			body := content[2:]
			// Reject non-UTF-8 metadata bodies (symmetric with the Rust impl,
			// which also rejects rather than lossily decoding); valid pigpen
			// content — markl IDs, blech32, text — is always UTF-8.
			if !utf8.ValidString(body) {
				return nil, errors.New("pigpen: metadata line body is not valid UTF-8")
			}
			doc.meta = append(doc.meta, metaLine{prefix: prefix, body: body})
		default:
			return nil, fmt.Errorf("%w: %q", errUnknownPrefix, prefix)
		}
	}
	if !closed {
		return nil, errNoClosingBoundary
	}

	// Anything after the closing boundary is the body, preceded by the
	// required blank-line separator.
	rest, _ := readAll(r)
	if len(rest) > 0 {
		if rest[0] != '\n' {
			return nil, errMissingSeparator
		}
		doc.body = rest[1:]
	}
	if doc.hasAtRef() && doc.hasInlineBody() {
		return nil, errAtRefWithInlineBody
	}
	return doc, nil
}

func readAll(r *bufio.Reader) ([]byte, error) {
	var b bytes.Buffer
	_, err := b.ReadFrom(r)
	return b.Bytes(), err
}
